### ISSUE TITLE
printer encoding CP437

### ISSUE THREAD

**Issue by mushunrek** [adafruit/Adafruit_CircuitPython_Thermal_Printer#24]
Hello everybody,

I am new here. I was working on a project using the Adafruit Mini Thermal Printer and I got into trouble when I wanted to use characters from extended ASCII tables. After looking into the user manual, it seems that the default encoding of the printer is CP437. So I was wondering, why one should use thas ASCII-encoding instead of the CP437-encoding in the _write_char funtion at line 193. Even though this encoding is a little out of date, it still manages most of the special characters from EU, additionally to the English characters. It would be nice to have some possibility to change the encoding, but I didn't exactly understand from the user manual how this works exactly.

If this is not the right place to give such proposition, I would be glad about a direction where to go instead :)

Bests,
mushunrek

**Comment by jposada202020** [comment]
If you have any questions related to the Adafruit Products, I would suggest to post it in the Adafruit Forums
https://forums.adafruit.com/, as you will have more visibility
Or the Adafruit Discord Channel, maybe some folks there have the answer already

I will keep this issue open for the moment to see if anybody familiar with this code could give you an answer.

Thanks

**Comment by mushunrek** [comment]
Thank you very much, but I opened the issue to make this proposition to change the code: for me the problem is solved, as I simply changed the function in my code. But I think that this change might be helpful: since no commands are sent through _write_char, and all other ASCII-characters are preserved in CP437, the functionality of the function will only be extended.

**Comment by jposada202020** [comment]
Oh I see thank you, would you mind doing a PR? :)

**Comment by mushunrek** [comment]
Hey! I am not yet super used to git as I am not a regular programmer. If I understand it correctly, I would need to create a new branch, make the change and do a pull request afterwards?
My solution back then was rather simplistic: as I needed to extend the Printerclass anyways, I simply rewrote the specific function. Here is the code:

```
ENCODING = "cp437"

class Printer(ThermalPrinter):
    def _write_char(self, char):
    # Write a single character to the printer.
        if char == "\r":
            return  # Strip carriage returns by skipping them.
        self._wait_timeout()
        self._uart.write(bytes(char, ENCODING))                                            # THIS IS THE ONLY CHANGE! [except for some differences in the variable names as I had to define a different __max_column...]
        delay = self._byte_delay_s
        # Add extra delay for newlines or moving past the last column.
        if char == "\n" or self._column == self.__max_column:
            if self._column == 0:
                # Feed line delay
                delay += (self._char_height + self._line_spacing) * self._dot_feed_s
            else:
                # Text line delay
                delay += (self._char_height * self._dot_print_s) + (
                    self._line_spacing * self._dot_feed_s
                )
            self._column = 0
        else:
            self._column += 1
        self._set_timeout(delay)
```
Since I am a newbie, it would be great if you could tell me how exactly to proceed. I could do a branch and then a pull request if that is the correct workflow.

If I am not mistaken, there is a second function [has_paper] that didn't work with my printer. But this should go into a different issue?

**Comment by tekktrik** [comment]
Resolved by #29
